# **An Exploratory Analysis of Payroll and Overtime Trends in Louisville Metro Government**

# **Step3-Exploratory Data Analysis (EDA)**

**Goal:** Exploratory Data Analysis (EDA) is conducted to understand the structure, distribution, and key characteristics of the employee salary dataset. This step helps identify patterns, trends, and variations in employee compensation, overtime payments, job roles, departments, and calendar years before proceeding to detailed visual and advanced analytical techniques.

The analysis includes univariate, bivariate, and multivariate approaches using descriptive statistics, groupby operations, pivot tables, and correlation analysis. These methods provide insights into salary distribution, workforce composition, departmental compensation differences, and relationships among numerical variables, forming a strong foundation for further analysis.

In [4]:
import pandas as pd # Import pandas library for data manipulation and analysis
import numpy as np  # Import numpy library for numerical computations and handling NaN values

pd.set_option('display.max_columns',None) # Set pandas option to display all columns in the DataFrame output
pd.set_option('display.float_format','{:.2f}'.format) # Set pandas option to format floating-point numbers to two decimal places

In [5]:
print('Python libraries imported successfully')

Python libraries imported successfully


In [6]:
df = pd.read_csv('/content/Louisville_Metro_KY_-_Employee_Salary_Data (1).csv') # Load the employee salary dataset into a Pandas DataFrame
print('Employee salary dataset loaded successfully')

Employee salary dataset loaded successfully


In [4]:
df.shape # Display the number of rows and columns in the dataset

(40816, 11)

# **3.1 Univariate Analysis**
Univariate analysis focuses on examining individual variables in the employee salary dataset—such as annual salary, overtime pay, departments, job titles, and calendar year—independently to understand their range, average values, frequency, and overall distribution.

# **(1) Department-wise Employee Count**

In [7]:
df['Department'].value_counts()

Department
Louisville Metro Police Department              7655
Louisville Fire                                 3037
Parks & Recreation                              2973
Public Works & Assets                           2705
Department of Corrections                       2565
                                                ... 
Office of Performance Improvement (inactive)       8
Internal Auditor                                   7
Office of Performance Improvement                  5
Office of Philanthropy                             4
PWA - Solid Waste Management Services              2
Name: count, Length: 80, dtype: int64

**Observation:**

The department-wise employee count reveals an uneven distribution of the workforce across the organization.

A few departments have significantly higher employee counts, indicating workforce concentration in key operational units.

Several departments have comparatively low employee representation, suggesting specialized or support-oriented functions.

The wide range between the most and least populated departments highlights structural differences in staffing requirements.

This imbalance reflects variations in departmental size, functional responsibilities, and organizational priorities.

**Overall Interpretation:**

Employee distribution across departments is highly uneven, with workforce concentration in select departments. This pattern suggests that operational demands and functional roles vary significantly across departments, influencing staffing levels and resource allocation within the organization.

# **(2)Job Title Frequency**

In [8]:
df['Job_Title'].value_counts()

Job_Title
Police Officer                    5045
Corrections Officer               1728
Police Sergeant                   1071
Firefighter 56hr                  1039
EMT                                654
                                  ... 
Compliance & Enforcement Supv        1
Mechanic II P/U-CDL                  1
Cashier Supvsr                       1
Lbr Rel Implementation Liaison       1
APCD Compliance & Enf Coord          1
Name: count, Length: 1526, dtype: int64

**Observation:**

The job title-wise analysis shows that the dataset includes a wide range of employee roles within the organization.

A small number of job titles account for a large proportion of employees, while many job titles have relatively low employee counts.

This indicates that operational and mid-level roles are more common, whereas senior or highly specialized positions are fewer in number.

Job titles with low frequency likely represent roles requiring specialized skills or higher levels of responsibility.

The uneven distribution of employees across job titles reflects the hierarchical structure and role-based workforce planning of the organization.

**Overall Interpretation:**

Employee distribution across job titles is uneven, with a strong concentration in operational roles that support day-to-day activities. In contrast, senior and specialized roles are limited, indicating a structured organizational hierarchy. These insights provide a foundation for further analysis of salary, overtime, and department-wise compensation patterns.

# **(3)Salary distribution across employees**

In [9]:
df['Annual_Rate'].describe()

count    40816.00
mean     60137.08
std      25604.50
min       1300.00
25%      43680.00
50%      56971.20
75%      74131.20
max     520000.00
Name: Annual_Rate, dtype: float64

**Observation:**

This summary provides key descriptive statistics for the Annual_Rate variable, including count, mean, standard deviation, minimum, maximum, and quartile values.

The difference between the mean and median (50%) indicates whether the salary distribution is symmetric or skewed.

The range between the minimum and maximum salaries highlights the spread of employee compensation levels.

The standard deviation reflects the degree of variation in annual salaries across employees.

**Overall Interpretation:**

The descriptive statistics show that employee salaries are not evenly distributed. Most employees earn within a moderate salary range, while a smaller number of high-paying roles increase the overall average, indicating a right-skewed salary distribution. This summary provides a foundational understanding of compensation patterns before further analysis.

# **3.2 Bivariate Analysis**
Bivariate analysis focuses on examining the relationship between two variables in the employee salary dataset—such as annual salary and department, annual salary and calender year, salary and years of service. This analysis helps understand how employee compensation varies across groups and how it changes over time or with experience.

# **1) Correlation Analysis of Annual_Rate and Overtime Pay**

In [10]:
df[['Annual_Rate', 'Overtime_Rate']].corr()

Annual_Rate  Overtime_Rate
Annual_Rate           1.00           0.21
Overtime_Rate         0.21           1.00

**Observation**

**The correlation between Annual_Rate and Overtime_Rate is 0.21, indicating a weak positive relationship**. This means that employees with higher annual salaries do not necessarily receive higher overtime rates. Overtime rates appear fairly similar across different salary levels.

**Overall Interpretation**

**Overtime rate is largely influenced by job role, department requirements, and operational needs rather than an employee’s base annual salary. As a result, increases in annual salary do not strongly translate into higher overtime rates**. This suggests that overtime pay is role-based rather than salary-driven, and base compensation and overtime compensation function independently in many cases.

# **2)Correlation Analysis of Annual Salary and YTD**

In [11]:
df[['Annual_Rate', 'YTD_Total']].corr()

Annual_Rate  YTD_Total
Annual_Rate         1.00       0.88
YTD_Total           0.88       1.00

**Observation :**
Correlation analysis shows that **Annual_Rate and YTD_Total** have a **strong positive relationship (correlation coefficient ≈ 0.88)**. This means employees with higher annual salaries consistently accumulate higher year‑to‑date earnings. The scatter of values aligns closely, indicating that base salary is the dominant driver of total compensation.

**Overall Interpretation:**  
Annual salary is the most reliable predictor of YTD earnings across the workforce. While overtime, incentives, and other pay components may add variability, the strong correlation **(close to 1.0)** confirms that base pay structure is the foundation of employee compensation. For portfolio presentation, this insight highlights that YTD totals are largely explained by annual salary, reinforcing the stability and predictability of salary‑driven earnings.

# **3) Annual Salary vs Year (CalYear)**

In [12]:
df.groupby('CalYear')['Annual_Rate'].describe()

count     mean      std     min      25%      50%      75%       max
CalYear                                                                       
2021    6667.00 50727.46 20104.09 1300.00 38043.20 50003.20 60840.00 216301.80
2022    6730.00 56314.67 21853.89 1300.00 43201.60 56409.60 68016.00 238277.00
2023    6985.00 59251.91 27591.01 1300.00 42120.00 55993.60 72779.20 520000.00
2024    7326.00 61731.79 25626.66 1300.00 45011.20 58656.00 76419.20 245425.31
2025    7198.00 65040.99 27223.94 1300.00 47278.40 62036.00 80246.40 350000.00
2026    5910.00 68201.44 26400.88 1300.00 50211.20 65291.20 81577.60 350000.00

**Observation:**

 Annual salaries rise steadily from **2021 to 2026**, with the **mean increasing from ~50,727 to ~68,201** and the **median following the same upward path**. **Minimum salaries stay fixed at 13,000, while maximum salaries fluctuate sharply, peaking at 520,000 in 2023 and 350,000 in 2025–2026**. **Standard deviation grows, showing widening pay variation.**

**Overall Interpretation:**
     Compensation trends point to systematic salary growth across the workforce, likely driven by policy or inflation adjustments. Rising mean and median confirm broad‑based increases, while spikes in maximum salaries highlight outliers that skew distribution. The data reflects both predictable upward salary trends and growing inequality in pay levels.

# **Reason**

- **Policy Adjustments:**
 Organizations often revise salary structures annually to align with government regulations, union agreements, or internal compensation policies. This systematic adjustment drives the steady rise in mean and median salaries.

- **Inflation:**
As the cost of living increases, salaries are raised to maintain employees’ purchasing power. This explains the consistent upward trend across the workforce.
- **Industry Growth & Talent Retention:**
 Expanding industries or competitive labor markets push companies to increase pay to attract and retain skilled employees.

- **Outliers in Maximum Salaries:**
 The sharp spikes in maximum salaries (e.g., 520,000 in 2023) are likely due to exceptional roles, executive packages, or one‑off contracts. These outliers skew the distribution and inflate the standard deviation.

- **Growing Inequality:**

   While the mean and median show broad‑based growth, the rising standard deviation indicates widening pay gaps. A small group of employees earns disproportionately high salaries compared to the majority, creating inequality.




**Summary**

The data reflects predictable salary growth across the workforce, driven by inflation and policy changes, while outliers in maximum pay highlight unequal distribution. In short, salaries are rising steadily for most employees, but the gap between the highest earners and the rest is widening.


# **4) Annual Salary vs Years of Service**

In [13]:
df['Years_of_Service'] = df['CalYear'] - df.groupby('Employee_Name')['CalYear'].transform('min')
df.groupby('Years_of_Service')['Annual_Rate'].describe()

count     mean      std     min      25%      50%  \
Years_of_Service                                                        
0                16573.00 51859.91 21247.55 1300.00 38708.80 50627.20   
1                 8415.00 58634.61 26954.92 1300.00 42848.00 55224.00   
2                 6456.00 65648.36 25021.20 1300.00 48755.20 63564.80   
3                 5073.00 71504.19 26289.45 1300.00 52852.80 70512.00   
4                 4052.00 73534.68 25838.11 1300.00 55328.00 73923.20   
5                  247.00 69398.04 25291.63 7367.36 53965.60 68286.40   

                      75%       max  
Years_of_Service                     
0                61880.00 350000.00  
1                71874.40 520000.00  
2                77869.80 245425.31  
3                84406.40 257696.58  
4                86008.00 257696.58  
5                81577.60 168368.39

**Observation:**
The analysis shows that employee salaries generally increase with years of service, particularly in the first four years. New hires start with a **mean salary of around 51,860, which rises steadily to over 73,500 by year four**. **Median values follow the same upward trend**, confirming that typical employees benefit from tenure. **At the same time, salary variability grows with experience, as seen in the rising standard deviation and the presence of very high outliers, which likely represent executives or specialized roles**. The dip in mean salary at year five appears to be an anomaly caused by the small sample size of employees in that category.

**Overall Interpretation:**

- **Positive Correlation:**

   Tenure is strongly linked to salary growth, especially in the first four years. This suggests structured pay progression policies or step increases early in careers.

- **Divergence in Long‑Term Outcomes:**

   While experience boosts pay, the widening salary range indicates that career trajectory, promotions, and role specialization drive long‑term differences.

- **Equity Considerations:**

   The presence of high outliers highlights opportunities for advancement, but also underscores that not all employees benefit equally from tenure.

- **Strategic Insight:**

   Organizations may want to examine whether salary growth beyond the early years is equitable and whether career pathways are transparent enough to ensure fair progression.


- **Why Salaries Rise with Tenure**

- **Structured Pay Progression:**
   Many organizations have salary bands or increments tied to years of service. Employees typically receive raises in the early years as part of probation completion, performance reviews, or standard progression policies.

- **Skill Development:**

   As employees gain experience, they become more valuable to the organization. This often translates into promotions or role upgrades, which bring higher pay.

- **Retention Strategy:**

  Employers raise salaries in the first few years to retain talent and reduce turnover. Early growth is a way to encourage employees to stay.
  Why Variability Increases Over Time

- **Career Trajectories Diverge:**

  After the initial years, employees move into different paths—some into management or specialized roles, others remain in standard positions. This creates a wider spread in salaries.

- **Outliers (Executives/Specialists):**

  Very high salaries among long‑tenured employees are usually linked to leadership positions or niche expertise. These outliers inflate the standard deviation.
- **Market Adjustments:**
   External factors like industry demand or competitive hiring can lead to uneven salary growth among experienced employees.


# **3.3 Multivariate Analysis**
Multivariate analysis involves analyzing more than two variables simultaneously to understand complex relationships within the employee salary dataset. This approach helps examine how multiple factors—such as annual salary, overtime pay, department, job title, and calendar year—interact with each other. Multivariate analysis provides deeper insights into combined effects, trends, and patterns that cannot be identified through univariate or bivariate analysis alone.

# **1) Average Salary & Overtime by Department and Year**

In [14]:
df.groupby(['Department', 'CalYear'])[['YTD_Total', 'Overtime_Rate']].mean()


YTD_Total  Overtime_Rate
Department                  CalYear                          
APCD                        2022      68009.17          23.32
                            2023      68985.31          42.89
                            2024      71391.07         121.75
                            2025      73010.04         138.12
                            2026      73444.76           0.00
...                                        ...            ...
Youth Transitional Services 2022      58125.59        4865.05
                            2023      58301.64        5830.55
                            2024      61185.92        4352.82
                            2025      66963.96        8453.43
                            2026      59670.87         368.00

[269 rows x 2 columns]

**Observation**

This analysis examines variations in average annual salary and overtime pay across different departments over multiple calendar years. The results show that salary and overtime levels differ significantly between departments. In several departments, both average salary and overtime pay display noticeable changes across years, indicating year-wise growth or fluctuation patterns.

**Overall Interpretation**

Employee compensation patterns vary by department and evolve over time. Departments with higher average salaries often reflect specialized roles or senior-level positions, while increased overtime pay in certain years suggests higher workload, staffing shortages, or operational demands. Overall, the combined analysis of salary and overtime by department and year provides insight into compensation structure, workforce utilization, and departmental resource allocation trends.

# **2) Average Salary by Department, Job Title, and Year**

In [15]:
df.groupby(['Department', 'Job_Title', 'CalYear'])['Annual_Rate'].mean()

Department                   Job_Title                    CalYear
APCD                         APCD Air Monitoring Chemist  2022      61734.40
                                                          2023      65083.20
                                                          2024      68848.00
                                                          2025      72300.80
                                                          2026      72300.80
                                                                      ...   
Youth Transitional Services  Transportation Supervisor    2022      55525.60
                                                          2023      68369.60
                                                          2024      70418.40
                                                          2025      73954.40
                                                          2026      73954.40
Name: Annual_Rate, Length: 7134, dtype: float64

**Observation**

The analysis shows significant variation in average annual salaries across different departments and job titles over multiple calendar years. Certain job roles consistently receive higher average salaries within specific departments, while other roles show gradual increases over time. Year-wise changes in salary levels indicate both steady growth in some roles and fluctuations in others, suggesting differences in career progression, departmental priorities, and workforce structure.

**Overall Interpretation**

Employee compensation is strongly influenced by a combination of department, job title, and year. Higher salaries are typically associated with senior or specialized roles and departments requiring advanced skills. Year-on-year salary changes reflect factors such as promotions, experience accumulation, policy revisions, and organizational growth. Overall, this multivariate analysis provides a comprehensive view of how role hierarchy and departmental context shape salary trends over time.

# **3 Year-wise Department Salary Variation Table Preparation**

In [16]:
df_filtered = df[df['CalYear'].between(2021, 2026)]
dept_yearwise = (
    df_filtered.groupby(['Department', 'CalYear'])['Annual_Rate']
    .mean()
    .reset_index()
)

dept_variation_table = dept_yearwise.pivot(
    index='Department',
    columns='CalYear',
    values='Annual_Rate'
)

dept_variation_table_reset = dept_variation_table.reset_index()

dept_variation_table_reset



CalYear                          Department     2021     2022     2023  \
0                                      APCD      NaN 66162.10 67700.63   
1                     Air Pollution Control 60635.59      NaN      NaN   
2                  Alcohol Beverage Control      NaN      NaN 52632.62   
3                Alcoholic Beverage Control      NaN      NaN      NaN   
4                       Belle of Louisville 30530.93 35550.62 34761.21   
..                                      ...      ...      ...      ...   
75                      Technology Services 74680.64      NaN      NaN   
76              Waterfront Development Corp 61933.16      NaN      NaN   
77       Waterfront Development Corporation      NaN 65523.47 66946.57   
78                 Youth Detention Services 53210.64      NaN      NaN   
79              Youth Transitional Services      NaN 51403.58 52300.21   

CalYear     2024     2025     2026  
0       70252.43 72853.40 73444.76  
1            NaN      NaN      NaN  
2            NaN      NaN      NaN  
3       58521.44 72724.13 72724.13  
4       38171.76 40506.96 42195.03  
..           ...      ...      ...  
75           NaN      NaN      NaN  
76           NaN      NaN      NaN  
77      72803.96 68056.12 68754.41  
78           NaN      NaN      NaN  
79      56637.93 58416.79 59302.87  

[80 rows x 7 columns]

# **Observation**

- In the department‑year salary pivot table, **NaN indicates that no salary data was available for that department in that year.**

- This could be due to restructuring, renaming, or missing records.

- After filtering the dataset for the years 2021 to 2026, annual salary values were grouped by Department and Calendar Year.

- The average annual salary for each department was calculated, and the results were reshaped into a pivot table:

- Rows → Departments
- Columns → Years (2021–2026)
- This structure helps to clearly view salary trends across departments year by year.

- # **Overall Interpretation**

- The department‑year variation table helps in understanding how average salary levels change over time within different departments.

- # **By comparing values across years, it becomes easier to identify:**

- Departments with consistent salary growth
- Departments showing sudden increases
- Departments with declining trends

- # **This analysis is useful for:**

- Evaluating compensation patterns
- Understanding departmental payroll changes
- Studying long‑term salary distribution across the organization

- # **Data Limitation**

- The 2026 data is not yet updated, which means insights for that year are incomplete.
- Any interpretation of 2026 values should therefore be approached cautiously until the dataset is refreshed with finalized records.


